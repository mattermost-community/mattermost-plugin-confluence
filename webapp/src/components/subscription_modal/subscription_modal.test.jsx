/*eslint max-nested-callbacks: ["error", 3]*/

import React from 'react';

import {shallow} from 'enzyme';

import Constants from '../../constants';

import SubscriptionModal from './subscription_modal';

describe('components/ChannelSettingsModal', () => {
    const baseProps = {
        theme: {},
        visibility: false,
        subscription: {},
        close: jest.fn(),
        saveChannelSubscription: jest.fn().mockResolvedValue({}),
        currentChannelID: 'abcabcabcabcabc',
        editChannelSubscription: jest.fn().mockResolvedValue({}),
    };

    test('subscription modal snapshot test', async () => {
        const props = {
            ...baseProps,
            visibility: true,
        };
        const wrapper = shallow(
            <SubscriptionModal {...props}/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('new space subscription', async () => {
        const props = {
            ...baseProps,
            visibility: true,
        };
        const wrapper = shallow(
            <SubscriptionModal {...props}/>,
        );
        wrapper.setState({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS,
            error: '',
            saving: false,
            pageID: '',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[0],
        });
        wrapper.instance().handleSubmit({preventDefault: jest.fn()});
        expect(wrapper.state().error).toBe('');
        expect(props.saveChannelSubscription).toHaveBeenCalledWith({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS.map((event) => event.value),
            channelID: 'abcabcabcabcabc',
            pageID: '',
            subscriptionType: 'space_subscription',
        });

        expect(props.editChannelSubscription).not.toHaveBeenCalled();
    });

    test('edit space subscription', async () => {
        const subscription = {
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS,
            pageID: '',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[0].value,
        };
        const props = {
            ...baseProps,
            subscription,
        };
        const wrapper = shallow(
            <SubscriptionModal {...props}/>,
        );
        wrapper.setState({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS,
            error: '',
            saving: false,
            pageID: '',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[0],
        });
        wrapper.instance().handleSubmit({preventDefault: jest.fn()});
        expect(wrapper.state().error).toBe('');
        expect(props.editChannelSubscription).toHaveBeenCalledWith({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS.map((event) => event.value),
            channelID: 'abcabcabcabcabc',
            pageID: '',
            subscriptionType: 'space_subscription',
        });
        expect(props.saveChannelSubscription).not.toHaveBeenCalled();
    });

    test('new page subscription', async () => {
        const props = {
            ...baseProps,
            visibility: true,
        };
        const wrapper = shallow(
            <SubscriptionModal {...props}/>,
        );
        wrapper.setState({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: '',
            events: Constants.CONFLUENCE_EVENTS,
            error: '',
            saving: false,
            pageID: '1234',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[1],
        });
        wrapper.instance().handleSubmit({preventDefault: jest.fn()});
        expect(wrapper.state().error).toBe('');
        expect(props.saveChannelSubscription).toHaveBeenCalledWith({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: '',
            events: Constants.CONFLUENCE_EVENTS.map((event) => event.value),
            channelID: 'abcabcabcabcabc',
            pageID: '1234',
            subscriptionType: 'page_subscription',
        });

        expect(props.editChannelSubscription).not.toHaveBeenCalled();
    });

    test('edit page subscription', async () => {
        const subscription = {
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS,
            pageID: '1234',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[1].value,
        };
        const props = {
            ...baseProps,
            subscription,
        };
        const wrapper = shallow(
            <SubscriptionModal {...props}/>,
        );
        wrapper.setState({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: '',
            events: Constants.CONFLUENCE_EVENTS,
            error: '',
            saving: false,
            pageID: '1234',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[1],
        });
        wrapper.instance().handleSubmit({preventDefault: jest.fn()});
        expect(wrapper.state().error).toBe('');
        expect(props.editChannelSubscription).toHaveBeenCalledWith({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: '',
            events: Constants.CONFLUENCE_EVENTS.map((event) => event.value),
            channelID: 'abcabcabcabcabc',
            pageID: '1234',
            subscriptionType: 'page_subscription',
        });
        expect(props.saveChannelSubscription).not.toHaveBeenCalled();
    });

    test('subscription data clean', async () => {
        const props = {
            ...baseProps,
            visibility: true,
        };
        const wrapper = shallow(
            <SubscriptionModal {...props}/>,
        );
        wrapper.setState({
            alias: '   Abc   ',
            baseURL: 'https://teST.com',
            spaceKey: 'test       ',
            events: Constants.CONFLUENCE_EVENTS,
            error: '',
            saving: false,
            pageID: '',
            subscriptionType: Constants.SUBSCRIPTION_TYPE[0],
        });
        wrapper.instance().handleSubmit({preventDefault: jest.fn()});
        expect(wrapper.state().error).toBe('');
        expect(props.saveChannelSubscription).toHaveBeenCalledWith({
            alias: 'Abc',
            baseURL: 'https://test.com',
            spaceKey: 'test',
            events: Constants.CONFLUENCE_EVENTS.map((event) => event.value),
            channelID: 'abcabcabcabcabc',
            pageID: '',
            subscriptionType: 'space_subscription',
        });

        expect(props.editChannelSubscription).not.toHaveBeenCalled();
    });
});
