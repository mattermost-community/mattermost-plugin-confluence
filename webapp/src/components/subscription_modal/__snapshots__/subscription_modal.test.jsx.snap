// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ChannelSettingsModal subscription modal snapshot test 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop="static"
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  onHide={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      Edit Your Confluence Subscription
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <div>
      <ConfluenceField
        fieldType="input"
        formControlStyle={Object {}}
        formGroupStyle={Object {}}
        label="Name"
        onChange={[Function]}
        placeholder="Enter a name for this subscription."
        readOnly={false}
        required={true}
        type="text"
        value=""
      />
      <ConfluenceField
        fieldType="input"
        formControlStyle={Object {}}
        formGroupStyle={Object {}}
        label="Confluence Base URL"
        onChange={[Function]}
        placeholder="Enter the Confluence Base URL."
        readOnly={false}
        required={true}
        type="text"
        value=""
      />
      <div>
        <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <ConfluenceField
            fieldType="dropDown"
            formControlStyle={Object {}}
            formGroupStyle={
              Object {
                "flex": "1",
                "marginRight": "20px",
              }
            }
            isMulti={false}
            isSearchable={false}
            label="Subscribe To"
            name="type"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "Space",
                  "value": "space_subscription",
                },
                Object {
                  "label": "Page",
                  "value": "page_subscription",
                },
              ]
            }
            readOnly={false}
            required={true}
            theme={Object {}}
            value={
              Object {
                "label": "Space",
                "value": "space_subscription",
              }
            }
          />
          <ConfluenceField
            fieldType="input"
            formControlStyle={
              Object {
                "height": "38px",
              }
            }
            formGroupStyle={
              Object {
                "flex": "1",
              }
            }
            label="Space Key"
            onChange={[Function]}
            placeholder="Enter the Confluence Space Key."
            readOnly={false}
            required={true}
            type="text"
            value=""
          />
        </div>
        <ConfluenceField
          fieldType="dropDown"
          formControlStyle={Object {}}
          formGroupStyle={Object {}}
          isMulti={true}
          label="Events"
          name="events"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "Comment Create",
                "value": "comment_created",
              },
              Object {
                "label": "Comment Update",
                "value": "comment_updated",
              },
              Object {
                "label": "Page Create",
                "value": "page_created",
              },
              Object {
                "label": "Page Update",
                "value": "page_updated",
              },
              Object {
                "label": "Page Trash",
                "value": "page_trashed",
              },
              Object {
                "label": "Page Restore",
                "value": "page_restored",
              },
              Object {
                "label": "Space Updated",
                "value": "space_updated",
              },
            ]
          }
          readOnly={false}
          required={true}
          theme={Object {}}
          value={
            Array [
              Object {
                "label": "Comment Create",
                "value": "comment_created",
              },
              Object {
                "label": "Comment Update",
                "value": "comment_updated",
              },
              Object {
                "label": "Page Create",
                "value": "page_created",
              },
              Object {
                "label": "Page Update",
                "value": "page_updated",
              },
              Object {
                "label": "Page Trash",
                "value": "page_trashed",
              },
              Object {
                "label": "Page Restore",
                "value": "page_restored",
              },
              Object {
                "label": "Space Updated",
                "value": "space_updated",
              },
            ]
          }
        />
      </div>
    </div>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      bsStyle="link"
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="primary"
    >
      Cancel
    </Button>
    <Button
      active={false}
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
      type="submit"
      variant="primary"
    >
      Save Subscription
    </Button>
  </ModalFooter>
</Modal>
`;
